### ISSUE TITLE
Determine confidence percentage for svm prediction

### ISSUE THREAD

**Issue by jeff1evesque** [jeff1evesque/machine-learning#1924]
Determine if confidence percentage can be returned for a given svm prediction.

**Comment by jeff1evesque** [comment]
This can be attained by setting the probability parameter to `True`, which will allow the use of the `predict_proba` method.

**Comment by jeff1evesque** [comment]
The following is an [example](http://stackoverflow.com/questions/21377699/scikit-learn-python-interpreting-svc-probabilities) implementation to compute the probability a target could be one of predefined class:

```python
>>> # example 1
...
>>> from sklearn import svm
>>> X = [[1,2,3], [2,3,4], [1,1,3], [2,3,5], [3,4,6], [2,3,4],[1,2,3]]
>>> y = ['cat', 'dog', 'cat','dog','fish','dog','cat']
>>> clf =svm.SVC(probability=True)
>>> clf.fit(X, y)
SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0, degree=3, gamma=0.0,
  kernel='rbf', max_iter=-1, probability=True, random_state=None,
  shrinking=True, tol=0.001, verbose=False)
>>> a=clf.decision_function([3,4,6])
>>> b=clf.predict_proba([3,4,6])
>>> c=clf.score(X, y)
>>>
>>> print clf.classes_
['cat' 'dog' 'fish']
>>> print clf.predict([3,4,6])
['fish']
>>> print  'predict', b
predict [[ 0.39247853  0.58474607  0.0227754 ]]
>>>
>>> # example 2
...
>>> d=clf.decision_function([1,2,3])
>>> e=clf.predict_proba([1,2,3])
>>> print clf.predict([1,2,3])
['cat']
>>> print  'predict', e
predict [[ 0.6131741   0.19514498  0.19168093]]
```

We notice that `# example 1` returned prediction of `fish`, though it had a probability of `0.0227754`, which is much lower than the probability for `dog` (0.58474607), or `cat` (0.39247853).  However, `# example 2` portrayed more accurate probability corresponding to its correct prediction of `cat`.

**Comment by jeff1evesque** [comment]
The following is an argument to implement `decision_function` instead of `predict_proba`:

>The SVC method decision_function gives per-class scores for each sample (or a single score per sample in the binary case). When the constructor option probability is set to True, class membership probability estimates (from the methods predict_proba and predict_log_proba) are enabled. In the binary case, the probabilities are calibrated using Platt scaling: logistic regression on the SVM’s scores, fit by an additional cross-validation on the training data. In the multiclass case, this is extended as per Wu et al. (2004).

>Needless to say, the cross-validation involved in Platt scaling is an expensive operation for large datasets. In addition, the probability estimates may be inconsistent with the scores, in the sense that the “argmax” of the scores may not be the argmax of the probabilities. (E.g., in binary classification, a sample may be labeled by predict as belonging to a class that has probability <½ according to predict_proba.) Platt’s method is also known to have theoretical issues. If confidence scores are required, but these do not have to be probabilities, then it is advisable to set probability=False and use decision_function instead of predict_proba.

With a supporting [article](http://www.csie.ntu.edu.tw/~cjlin/papers/svmprob/svmprob.pdf) reference by *Wu, Lin and Weng*.

**Comment by jeff1evesque** [comment]
We will continue to implement the `predict_proba` method.  However, we will implement the following:

- [brier score loss](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.brier_score_loss.html#sklearn.metrics.brier_score_loss)
- [probability calibration](http://scikit-learn.org/stable/modules/calibration.html#calibration)

**Comment by jeff1evesque** [comment]
Actually, for completeness, we will implement both `predict_proba`, as well as the `decision_function`.  The former attempts to calculate the probability a given point `a(n)` is of one of defined classifiers, while the latter calculates the given point `a(n)` to the defined hyperplane(s) for each of the defined classifiers.  By implementing both solutions, the user will be able to determine which measure is most useful.  It is also important to note, that our [above cases](https://github.com/jeff1evesque/machine-learning/issues/1924#issuecomment-159491052) were poor examples of both implementations, since the given dataset is very small to make such confidence related measurements.

**Comment by jeff1evesque** [comment]
47f187c: defining `probability=True` in `clf = svm.SVC(kernel=kernel_type, probability=True)` is required, when implementing the `predict_proba` method.

**Comment by jeff1evesque** [comment]
42818cd: ensuring corresponding structures are `list` type, ensure it can be json serialized.

**Comment by jeff1evesque** [comment]
We need to ensure our float validator converts input values to string, prior to validation.

**Comment by jeff1evesque** [comment]
We need to ensure that `valid_float.js` will validate values such as `2.932604439796274e-8` as a float.  Otherwise, `confidence.probability.every(checkValidFloat)` will fail from `result_display.jsx`:

```javascript
...
        if (
            serverResult &&
            serverResult.model &&
            serverResult.model == 'svm' &&
            confidence.classes &&
            confidence.classes.every(checkValidFloat) &&
            confidence.probability &&
            confidence.probability.every(checkValidFloat) &&
            confidence.decision_function &&
            confidence.decision_function.every(checkValidFloat)
        ) {
            adjustedConfidence = {
                'classes': confidence.classes.join(', '),
                'probability': confidence.probability.join(', '),
                'decision-function': confidence.decision_function.join(', ')
            }
        }
...
```

**Comment by jeff1evesque** [comment]
Per defined linters, we need to ensure various files have less than 80 characters per line.
